File: leetcode_0141/cpp/leetcode_0141.cpp
//
// Created by wzy on 2022/3/22.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>

using namespace std;

struct ListNode{
    int val;
    ListNode *next;
    ListNode() : val(0),next(nullptr) {}
    ListNode(int x) : val(x),next(nullptr) {}
    ListNode(int x,ListNode *next):val(x),next(next) {}
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(!head)
            return false;
        //快慢双指针 是否重合
        ListNode* fast = head->next;
        ListNode* slow = head;
        while(fast && fast->next){
            if(fast == slow)
                return true;
            fast = fast->next->next;
            slow = slow->next;
        }
        return false;
    }
};

void delList(ListNode* head){
    ListNode* cur = head;
    while(cur && cur->next){
        ListNode* next = cur->next;
        cur = nullptr;
        cur = next;
    }
    cur = nullptr;
}

template<typename T>
void printList(std::list<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
std::list<T> toList(ListNode* head){
    std::list<T> l;
    ListNode* cur = head;
    while(cur){
        l.push_back(cur->val);
        cur = cur->next;
    }
    return l;
}

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main() {
//    ListNode* head = new ListNode(3);
//    ListNode* h1 = new ListNode(2);
//    ListNode* h2 = new ListNode(0);
//    ListNode* h3 = new ListNode(-4);
//    head->next = h1;h1->next = h2;h2->next = h3;
//    h3->next = h1;

    ListNode* head = new ListNode(1);
    cout<<boolalpha<<Solution().hasCycle(head)<<endl;
}

File: leetcode_0141/java/leetcode_0141.java
import java.util.LinkedList;
import java.util.List;

public class leetcode_0141 {
    public class ListNode<T>{
        T val;
        ListNode next;
        ListNode(){}
        ListNode(T val) {this.val = val;}
        ListNode(T val, ListNode next) {this.val = val;this.next = next;}
    }
    public class Solution {
        public boolean hasCycle(ListNode head) {
            if(head == null)
                return false;
            //快慢双指针 是否重合
            ListNode fast = head.next;
            ListNode slow = head;
            while(fast != null && fast.next != null){
                if(fast == slow)
                    return true;
                fast = fast.next.next;
                slow = slow.next;
            }
            return false;
        }
        public <T> void printVec(List<T> list){
            String res = "[";
            for(T ele : list){
                res += ele+",";
            }
            res = res.substring(0,res.length() - 1);
            res += "]";
            System.out.println(res);
        }

        public <T> List<T> toList(ListNode head){
            List<T> list = new LinkedList<T>();
            ListNode cur = head;
            while(cur != null){
                list.add((T) cur.val);
                cur = cur.next;
            }
            return list;
        }
    }
    public static void main(String[] args) {
        Solution solution = new leetcode_0141().new Solution();
//        ListNode head = new leetcode_0141().new ListNode(3);
//        ListNode h1 = new leetcode_0141().new ListNode(2);
//        ListNode h2 = new leetcode_0141().new ListNode(0);
//        ListNode h3 = new leetcode_0141().new ListNode(-4);
//        head.next = h1;h1.next = h2;h2.next = h3;h3.next = h1;
        ListNode head = new leetcode_0141().new ListNode(1);
        System.out.println(solution.hasCycle(head));
    }
}
